src/main/java/com/google/devtools/build/lib/skyframe/SequencedSkyframeExecutor.java

```diff
@@ -376,13 +376,18 @@ private WorkspaceInfoFromDiff handleDiffs(
     int fsvcThreads = buildRequestOptions == null ? 200 : buildRequestOptions.fsvcThreads;
     handleDiffsWithCompleteDiffInformation(
         tsgm, modifiedFilesByPathEntry, managedDirectoriesChanged, fsvcThreads);
-    handleDiffsWithMissingDiffInformation(
-        eventHandler,
-        tsgm,
-        pathEntriesWithoutDiffInformation,
-        checkOutputFiles,
-        managedDirectoriesChanged,
-        fsvcThreads);
+    if (checkOutputFiles || checkExternalFiles || !pathEntriesWithoutDiffInformation.isEmpty()) {
```

Comment:
Sorry I didn't notice this before:

Unfortunately the github PR interface doesn't let me comment on unchanged lines, but I'm a bit concerned about lines 493-498 in this PR: the guard of when to check inside handleDiffs seems to be both stricter and looser than this check. In substance, this flag seems to be inconsistently assuming that "external repository" files should count as "external" files. Do you want them to? I would naively think that you would want this flag only for external non-repository files. I don't think it should apply to both non-repository and repository files, we'd probably want separate flags. Here are the inconsistencies I see:

(1) This check doesn't respect the value of customDirtinessCheckers. I think the only place that's set is in BazelRepositoryModule, with a RepositoryDirectoryDirtinessChecker. Based on that class's javadoc, it looks like this check would lead to a bug where a `--nofetch` build followed by a `--fetch` build didn't properly update repositories when running with `--watchfs`, `--noexperimental_check_external_files` and `--noexperimental_check_output_files`.

(2) Similarly, it looks like *either* checkExternalFiles should make us ignore externalFilesKnowledge.anyFilesInExternalRepoSeen or it should be overridden. Right now, there's a redundant check, which I think you can exercise by doing a `--nowatchfs build --nobuild` on a project with a file in an external repository, then on an incremental build do `--nowatchfs build --nocheck_external_files` , and having a file in an external repository. That will do a full filesystem scan. However, running with `--watchfs` (and `--noexperimental_check_output_files`) will skip the full filesystem scan. That seems wrong. Also, anyFIlesInExternalRepoSeen can refer either to "normal" external repository files (which maybe you want to suppress checking for) and to "managed directories" (see ExternalFilesHelper) which is a pretty tricky concept. Is it actually the case that you want to disable checks in managed directories with this flag, even if you do want to suppress repository checks? I think probably not, although I'm not positive when managed directories are expected to be updated by the corresponding repository rules. If you don't want to disable such checks, you'll probably need to split the ExternalFilesKnowledge boolean into two, and track them separately in ExternalFilesHelper.

Reply:
Thank you for the catch!

> In substance, this flag seems to be inconsistently assuming that "external repository" files should count as "external" files. Do you want them to? 

I don't know the difference. In my case I want to not scan any files which a normal users would think are owned by bazel. In my case, that is a bunch of `http_archive`s and packages registered by `rules_python` and `rules_nodejs`.  Basically things owned by repository rules which I don't manage the files for. I'm assuming that is everything in the `externals/` folder? It was surprising to me that these needed to be scanned for changes on every invocation (especially in CI where my users don't have direct filesystem writing and where doing anything outside of the bazel sandbox is very unorthodox).

Thanks for those extra cases! I'll add tests for those and look at updating this logic to match the same as are on 493-498. I'm assuming I'd want to check my flag in that section too just like `checkOutputFiles` is used there

Reply:
I added a test for (1) and it seems to pass. Maybe I didn't understand the case you were trying to get at:
```
  test_no_fetch_then_fetch() {
    setup_remote
    bazel build \
      --nofetch \
      --noexperimental_check_external_files \
      --noexperimental_check_output_files \
      --watchfs \
      @remote//:g >& $TEST_log && fail "Expected build to fail" || true
    expect_log "no such package '@bazel_tools//tools/build_defs/repo'"
    bazel build \
      --fetch \
      --noexperimental_check_external_files \
      --noexperimental_check_output_files \
      --watchfs \
      @remote//:g >& $TEST_log || fail "Expected build to pass"
  }
```

Reply:
And for (2) I didn't really follow. I tried adding a test which I think did as you said but the second build still works. Did you want me to edit the filesystem in the meantime and show that the scan didn't happen? Or did you think it should have done a scan for that?

Reply:
Just doing a normal build from my tests there is a `com.google.devtools.build.lib.rules.repository.RepositoryDirectoryDirtinessChecker@6c76d4a0` in that variable. Can I short circuit that if my flag is set?